### PR TITLE
feat: template for blue/green deployments

### PR DESCRIPTION
**Pull request by jmpalomares**
https://parcellab.atlassian.net/browse/INF-589
## Description
This Pull Request aims to seamlessly integrate blue-green deployments into the Argo Rollouts platform, enhancing its capabilities for continuous delivery and ensuring minimal downtime during application updates. Leveraging Argo's robust Kubernetes controller, this integration enables organisations to orchestrate smooth transitions between different versions of their applications while maintaining uninterrupted service for end-users.

Key Features:

- Automated Traffic Shifting: Incrementally shifts traffic from the active environment (blue) to the new one (green), facilitating real-time monitoring and instant rollback.
- Health-Based Promotion: Validates the stability of the new version before promoting it to active status, ensuring user experience remains consistent.
- Rollback Mechanism: Provides a swift rollback option in case of complications, minimising downtime and mitigating potential business impacts.
- BlueGreen Deployment Strategy: Allows users to reduce the amount of time multiple versions running at the same time.
- Canary Deployment Support: A canary rollout is a deployment strategy where the operator releases a new version of their application to a small percentage of the production traffic.
- Customisation and Extensibility: Offers extensive Customisation options to tailor deployment strategies to specific requirements.

This integration enhances Argo Rollouts' versatility and reliability, empowering organisations to streamline their release processes and enhance confidence in their deployment pipelines.

https://argo-rollouts.readthedocs.io/en/stable/

## Checklist

- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

### CODE REVIEW

File: parcellab/common/templates/_rolloutdeployment.tpl [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+apiVersion: argoproj.io/v1alpha1
+kind: AnalysisTemplate
+metadata:
+  name: {{ $metric.name }}-analysis
+spec:
+  args:
+  - name: {{ $metric.name }}
+  metrics:
+{{ toYaml $metric | indent 4 }}
+
```
Comment (andibeuge) [inline parcellab/common/templates/_rolloutdeployment.tpl:68]:
```suggestion
---
```

File: parcellab/common/templates/_rolloutdeployment.tpl [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+{{- if ne (add $index 1) (len $rolloutSpec.metrics) }}
```
Comment (andibeuge) [inline parcellab/common/templates/_rolloutdeployment.tpl:69]:
```suggestion
```

File: parcellab/common/templates/_rolloutdeployment.tpl [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+---
```
Comment (andibeuge) [inline parcellab/common/templates/_rolloutdeployment.tpl:70]:
```suggestion
```

File: parcellab/common/templates/_rolloutdeployment.tpl [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+{{- end }}
+{{- end }}
+{{- end }}
+---
```
Comment (andibeuge) [inline parcellab/common/templates/_rolloutdeployment.tpl:74]:
```suggestion
```

File: parcellab/common/templates/_rolloutdeployment.tpl [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+{{- end }}
+{{- if .Values.rolloutDeployment.canary }}
+  {{- $rolloutSpec = .Values.rolloutDeployment.canary }}
+{{- end }}
+{{- if $rolloutSpec.metrics }}
+{{- range $index, $metric := $rolloutSpec.metrics }}
+apiVersion: argoproj.io/v1alpha1
+kind: AnalysisTemplate
+metadata:
+  name: {{ $metric.name }}-analysis
```
Comment (andibeuge) [inline parcellab/common/templates/_rolloutdeployment.tpl:62]: Let's add the $name here as well to ensure resource name uniquness

File: parcellab/microservice/values.yaml [hunk 6]
```diff
@@ -71,6 +71,49 @@ externalSecret:
...
+  #     successCondition: result <= 0.01
+  #     failureLimit: 3
+  #     provider:
+  #       datadog:
+  #         apiVersion: v2
+  #         interval: 5m
+  #         query: |
+  #           sum:requests.error.rate{service:app}
+  #   spec:
+  #     activeService: service-name
```
Comment (andibeuge) [inline parcellab/microservice/values.yaml:108]: Would it make sense to generate this name out of the deployment etc ?

Comment (jmpalomares) [reply parcellab/microservice/values.yaml:108]: that name is generated from the deployment already yes

Comment (jmpalomares) [reply parcellab/microservice/values.yaml:108]: that is already generated from the deployment and is integrated in the template. I will remove it from the examples.

**Review by andibeuge — APPROVED** [review]

**Pull request merged**